File: re/mame/lua/input.lua
-- Controller discovery and deterministic button schedules.

return function(options)
    local core = options.core
    local machine = core.machine
    local json_string = core.json_string
    local json_array = core.json_array
    local json_object = core.json_object
    local current_frame = options.current_frame
    local root = options.root or "."
    local mode = (os.getenv("OPENALADDIN_INPUT_MODE") or "inject"):lower()
    if mode ~= "inject" and mode ~= "record" and mode ~= "playback" then
        error("OPENALADDIN_INPUT_MODE must be inject, record, or playback")
    end
    local write_input = options.write_input

    local function split(value, separator)
        local result = {}
        for item in (value .. separator):gmatch("(.-)" .. separator) do
            result[#result + 1] = item
        end
        return result
    end

    local fields = {}
    local field_names = {}
    local controller_port = machine.ioport.ports[":ctrl1:mdpad:PAD"]
    for port_tag, port in pairs(machine.ioport.ports) do
        for field_name, field in pairs(port.fields) do
            local normalized = field_name:lower()
            if normalized:find("p1", 1, true)
                or normalized:find("player 1", 1, true)
                or normalized:find("1p", 1, true) then
                fields[#fields + 1] = {
                    name = field_name,
                    normalized = normalized,
                    field = field,
                    port = port_tag
                }
                field_names[#field_names + 1] = json_string(field_name)
            end
        end
    end
    table.sort(fields, function(left, right) return left.name < right.name end)
    table.sort(field_names)

    local function find_field(words)
        for _, entry in ipairs(fields) do
            local matches = true
            for _, word in ipairs(words) do
                if not entry.normalized:find(word, 1, true) then
                    matches = false
                    break
                end
            end
            if matches then return entry.field end
        end
        return nil
    end

    local function find_field_any(candidates)
        for _, words in ipairs(candidates) do
            local field = find_field(words)
            if field then return field end
        end
        return nil
    end

    local controls = {
        up = find_field({ "up" }),
        down = find_field({ "down" }),
        left = find_field({ "left" }),
        right = find_field({ "right" }),
        a = find_field_any({ { "button 1" }, { "p1 a" }, { "player 1 a" } }),
        b = find_field_any({ { "button 2" }, { "p1 b" }, { "player 1 b" } }),
        c = find_field_any({ { "button 3" }, { "p1 c" }, { "player 1 c" } }),
        start = find_field({ "start" })
    }

    -- MAME exposes the Mega Drive pad as an active-low eight-bit port.  The
    -- input-run format deliberately exposes the same logical order as the
    -- Genesis controller, but with active-high bits so it is readable and
    -- portable across clients.
    local button_order = { "up", "down", "left", "right", "a", "b", "c", "start" }
    local button_masks = {}
    for index, name in ipairs(button_order) do
        button_masks[name] = 1 << (index - 1)
    end

    -- The Mega Drive input port is not laid out in canonical button order:
    -- B is bit 4, C is bit 5, and A is bit 6. Keep the portable timeline in
    -- Genesis order while deriving the physical masks from MAME's fields.
    local port_masks = {}
    for _, name in ipairs(button_order) do
        local field = controls[name]
        port_masks[name] = field and field.mask or button_masks[name]
    end

    local function canonical_mask()
        local raw = controller_port and controller_port:read() or 0xff
        local mask = 0
        for _, name in ipairs(button_order) do
            local port_mask = port_masks[name]
            if port_mask and (raw & port_mask) == 0 then
                mask = mask | button_masks[name]
            end
        end
        return mask
    end

    local function token_for_mask(mask)
        local pressed = {}
        for index, name in ipairs(button_order) do
            if (mask & button_masks[name]) ~= 0 then
                pressed[#pressed + 1] = name
            end
        end
        return #pressed == 0 and "none" or table.concat(pressed, "+")
    end

    if write_input then
        write_input({
            { "type", json_string("header") },
            { "format", json_string("openaladdin-input-v1") },
            { "controller_mapping", json_string("mame-genesis-3button-v1") },
            { "buttons", json_array((function ()
                local values = {}
                for index, name in ipairs(button_order) do
                    values[index] = json_string(name)
                end
                return values
            end)()) },
            { "mask_bits", json_object((function ()
                local values = {}
                for _, name in ipairs(button_order) do
                    values[#values + 1] = { name, tostring(button_masks[name]) }
                end
                return values
            end)()) },
            { "frame_semantics", json_string(
                "I[N] is the controller input used for the transition S[N] -> S[N+1] at synchronization boundary N"
            ) },
            { "frame_contract", json_string("S[N] = synchronized state at boundary N; I[N] = input for S[N] -> S[N+1]") },
            { "mode", json_string(mode) }
        })
    end

    local function expand_schedule(value)
        local result = {}
        for _, item in ipairs(split(value, ",")) do
            local token, count = item:match("^%s*(.-)%s*[*:](%d+)%s*$")
            if token then
                for _ = 1, tonumber(count) do result[#result + 1] = token:lower() end
            else
                result[#result + 1] = item:lower():gsub("^%s+", ""):gsub("%s+$", "")
            end
        end
        return result
    end

    local input = {}
    local schedule_tokens = expand_schedule(os.getenv("OPENALADDIN_INPUT") or "none")

    local function press_token(token)
        if token == "" or token == "none" then return end
        for part in token:gmatch("[^+]+") do
            local field = controls[part]
            if field then field:set_value(1) end
        end
    end

    local experiment = dofile(root .. "/re/mame/lua/experiment.lua")({
        core = core,
        write_record = options.write_record,
        write_state = options.write_state,
        current_frame = current_frame,
        press_token = press_token,
        expand_schedule = expand_schedule
    })

    local function clear()
        for _, entry in ipairs(fields) do entry.field:set_value(0) end
    end

    function input.apply()
        if mode == "record" or mode == "playback" then
            local mask = canonical_mask()
            local token = token_for_mask(mask)
            if write_input then
                write_input({
                    { "frame", tostring(current_frame()) },
                    { "mask", tostring(mask) },
                    { "buttons", json_array((function ()
                        local values = {}
                        for index, name in ipairs(button_order) do
                            if (mask & button_masks[name]) ~= 0 then
                                values[#values + 1] = json_string(name)
                            end
                        end
                        return values
                    end)()) }
                })
            end
            return token
        end

        clear()
        if experiment.has_actions() then return experiment.tick() end
        local token = schedule_tokens[current_frame() + 1] or "none"
        press_token(token)
        return token
    end

    function input.fields_json()
        return "[" .. table.concat(field_names, ",") .. "]"
    end

    function input.controller_value()
        return controller_port and controller_port:read() or 0
    end

    function input.canonical_mask()
        return canonical_mask()
    end

    function input.action_spec()
        return experiment.action_spec()
    end

    return input
end

File: genie/platforms/genesis/input.py
"""Genesis three-button controller mapping and input-token helpers."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


INPUT_BUTTONS = ("up", "down", "left", "right", "a", "b", "c", "start")
INPUT_MASKS = {name: 1 << index for index, name in enumerate(INPUT_BUTTONS)}
INPUT_MAPPING = "mame-genesis-3button-v1"

# Early recordings labelled the raw Mega Drive bits linearly as A/B/C. MAME's
# physical fields are B/C/A at those bit positions.
LEGACY_BUTTON_REMAP = {"a": "b", "b": "c", "c": "a"}


@dataclass(frozen=True, slots=True)
class Genesis3ButtonInput:
    """Portable logical representation of a Genesis three-button pad."""

    buttons: tuple[str, ...] = INPUT_BUTTONS
    masks: dict[str, int] | None = None
    mapping: str = INPUT_MAPPING

    def __post_init__(self) -> None:
        if self.masks is None:
            object.__setattr__(
                self,
                "masks",
                {name: 1 << index for index, name in enumerate(self.buttons)},
            )

    def buttons_for_mask(self, mask: int) -> list[str]:
        if mask < 0 or mask > 0xFF:
            raise SystemExit(f"input mask out of range: {mask}")
        assert self.masks is not None
        return [name for name in self.buttons if mask & self.masks[name]]

    def token_for_mask(self, mask: int) -> str:
        buttons = self.buttons_for_mask(mask)
        return "+".join(buttons) if buttons else "none"

    def client_input_tokens(
        self,
        header: dict[str, Any] | None,
        tokens: list[str],
    ) -> list[str]:
        """Translate recordings made before the controller mapping marker."""

        if header and header.get("controller_mapping") is not None:
            return list(tokens)
        return [
            "+".join(LEGACY_BUTTON_REMAP.get(part, part) for part in token.split("+"))
            if token != "none" else token
            for token in tokens
        ]


GENESIS_3BUTTON_INPUT = Genesis3ButtonInput()


def buttons_for_mask(mask: int) -> list[str]:
    return GENESIS_3BUTTON_INPUT.buttons_for_mask(mask)


def token_for_mask(mask: int) -> str:
    return GENESIS_3BUTTON_INPUT.token_for_mask(mask)


def client_input_tokens(
    header: dict[str, Any] | None,
    tokens: list[str],
) -> list[str]:
    return GENESIS_3BUTTON_INPUT.client_input_tokens(header, tokens)


__all__ = [
    "Genesis3ButtonInput",
    "GENESIS_3BUTTON_INPUT",
    "INPUT_BUTTONS",
    "INPUT_MASKS",
    "INPUT_MAPPING",
    "LEGACY_BUTTON_REMAP",
    "buttons_for_mask",
    "client_input_tokens",
    "token_for_mask",
]
